// @opengsn/gsn/dist/src/cli/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// TODO: allow reading network URLs from 'truffle-config.js'
const commander_1 = __importDefault(require("commander"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
exports.networks = new Map([
    ['localhost', 'http://127.0.0.1:8545'],
    ['xdai', 'https://dai.poa.network'],
    ['ropsten', 'https://ropsten.infura.io/v3/c3422181d0594697a38defe7706a1e5b'],
    ['rinkeby', 'https://rinkeby.infura.io/v3/c3422181d0594697a38defe7706a1e5b'],
    ['kovan', 'https://kovan.infura.io/v3/c3422181d0594697a38defe7706a1e5b'],
    ['mainnet', 'https://mainnet.infura.io/v3/c3422181d0594697a38defe7706a1e5b']
]);
function supportedNetworks() {
    return Array.from(exports.networks.keys());
}
exports.supportedNetworks = supportedNetworks;
function getNetworkUrl(network = '') {
    var _a, _b;
    const match = (_a = network.match(/^(https?:\/\/.*)/)) !== null && _a !== void 0 ? _a : [];
    return (_b = exports.networks.get(network)) !== null && _b !== void 0 ? _b : match[0];
}
exports.getNetworkUrl = getNetworkUrl;
function getMnemonic(mnemonicFile) {
    if (mnemonicFile == null) {
        return;
    }
    console.log('Using mnemonic from file ' + mnemonicFile);
    return fs_1.default.readFileSync(mnemonicFile, { encoding: 'utf8' }).replace(/\r?\n|\r/g, '');
}
exports.getMnemonic = getMnemonic;
function getRelayHubConfiguration(configFile) {
    if (configFile == null) {
        return;
    }
    console.log('Using hub config from file ' + configFile);
    const file = fs_1.default.readFileSync(configFile, { encoding: 'utf8' });
    return JSON.parse(file);
}
exports.getRelayHubConfiguration = getRelayHubConfiguration;
function getPaymasterAddress(paymaster) {
    return getAddressFromFile('build/gsn/Paymaster.json', paymaster);
}
exports.getPaymasterAddress = getPaymasterAddress;
function getRelayHubAddress(defaultAddress) {
    return getAddressFromFile('build/gsn/RelayHub.json', defaultAddress);
}
exports.getRelayHubAddress = getRelayHubAddress;
function getAddressFromFile(path, defaultAddress) {
    if (defaultAddress == null) {
        if (fs_1.default.existsSync(path)) {
            const relayHubDeployInfo = fs_1.default.readFileSync(path).toString();
            return JSON.parse(relayHubDeployInfo).address;
        }
    }
    return defaultAddress;
}
function saveContractToFile(address, workdir, filename) {
    fs_1.default.mkdirSync(workdir, { recursive: true });
    fs_1.default.writeFileSync(path_1.default.join(workdir, filename), `{ "address": "${address}" }`);
}
function saveDeployment(deploymentResult, workdir) {
    saveContractToFile(deploymentResult.stakeManagerAddress, workdir, 'StakeManager.json');
    saveContractToFile(deploymentResult.penalizerAddress, workdir, 'Penalizer.json');
    saveContractToFile(deploymentResult.relayHubAddress, workdir, 'RelayHub.json');
    saveContractToFile(deploymentResult.naivePaymasterAddress, workdir, 'Paymaster.json');
    saveContractToFile(deploymentResult.forwarderAddress, workdir, 'Forwarder.json');
}
exports.saveDeployment = saveDeployment;
function showDeployment(deploymentResult, title, paymasterTitle = undefined) {
    if (title != null) {
        console.log(title);
    }
    console.log(`
  RelayHub: ${deploymentResult.relayHubAddress}
  StakeManager: ${deploymentResult.stakeManagerAddress}
  Penalizer: ${deploymentResult.penalizerAddress}
  Forwarder: ${deploymentResult.forwarderAddress}
  Paymaster ${paymasterTitle != null ? '(' + paymasterTitle + ')' : ''}: ${deploymentResult.naivePaymasterAddress}`);
}
exports.showDeployment = showDeployment;
function loadDeployment(workdir) {
    function getAddress(name) {
        const address = getAddressFromFile(path_1.default.join(workdir, name + '.json'));
        if (address != null) {
            return address;
        }
        throw new Error('no address for ' + name);
    }
    return {
        relayHubAddress: getAddress('RelayHub'),
        stakeManagerAddress: getAddress('StakeManager'),
        penalizerAddress: getAddress('Penalizer.json'),
        forwarderAddress: getAddress('Forwarder.json'),
        naivePaymasterAddress: getAddress('Paymaster.json')
    };
}
exports.loadDeployment = loadDeployment;
function gsnCommander(options) {
    options.forEach(option => {
        switch (option) {
            case 'n':
                commander_1.default.option('-n, --network <url|name>', 'network name or URL to an Ethereum node', 'localhost');
                break;
            case 'f':
                commander_1.default.option('-f, --from <address>', 'account to send transactions from (default: the first account with balance)');
                break;
            case 'h':
                commander_1.default.option('-h, --hub <address>', 'address of the hub contract (default: the address from build/gsn/RelayHub.json if exists)');
                break;
            case 'm':
                commander_1.default.option('-m, --mnemonic <mnemonic>', 'mnemonic file to generate private key for account \'from\' (default: empty)');
                break;
            case 'g':
                commander_1.default.option('-g, --gasPrice <number>', 'gas price to give to the transaction. Defaults to 1 gwei.', '1000000000');
                break;
        }
    });
    return commander_1.default;
}
exports.gsnCommander = gsnCommander;
