// @opengsn/gsn/dist/test/dummies/BadContractInteractor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ContractInteractor_1 = __importDefault(require("../../src/relayclient/ContractInteractor"));
class BadContractInteractor extends ContractInteractor_1.default {
    constructor(provider, config, failValidateARC) {
        super(provider, config);
        this.failValidateARC = failValidateARC;
    }
    async validateAcceptRelayCall(relayRequest, signature, approvalData) {
        if (this.failValidateARC) {
            return {
                paymasterAccepted: false,
                reverted: true,
                returnValue: BadContractInteractor.message
            };
        }
        return await super.validateAcceptRelayCall(relayRequest, signature, approvalData);
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    async sendSignedTransaction(rawTx) {
        throw new Error(BadContractInteractor.wrongNonceMessage);
    }
}
exports.default = BadContractInteractor;
BadContractInteractor.message = 'This is not the contract you are looking for';
BadContractInteractor.wrongNonceMessage = 'the tx doesn\'t have the correct nonce';
